File: frontend/src/services/postService.js
import apiClient from './apiClient'

const postService = {
  /** Fetch paginated, sorted list of all posts */
  getAllPosts: async ({ page = 0, size = 10, sortBy = 'createdAt', sortDir = 'desc' } = {}) => {
    const response = await apiClient.get('/posts', { params: { page, size, sortBy, sortDir } })
    return response.data
  },

  /** Fetch a single post by ID */
  getPostById: async (postId) => {
    const response = await apiClient.get(`/posts/${postId}`)
    return response.data
  },

  /** Create a new post */
  createPost: async (data) => {
    const response = await apiClient.post('/posts', data)
    return response.data
  },

  /** Update an existing post */
  updatePost: async (postId, data) => {
    const response = await apiClient.put(`/posts/${postId}`, data)
    return response.data
  },

  /** Delete a post */
  deletePost: async (postId) => {
    await apiClient.delete(`/posts/${postId}`)
  },

  /** Toggle like/unlike on a post */
  toggleLike: async (postId) => {
    const response = await apiClient.post(`/posts/${postId}/like`)
    return response.data
  },

  /** Full-text search */
  searchPosts: async (query, page = 0, size = 10) => {
    const response = await apiClient.get('/posts/search', { params: { query, page, size } })
    return response.data
  },

  /** Filter posts by category */
  getPostsByCategory: async (category, page = 0, size = 10) => {
    const response = await apiClient.get(`/posts/category/${category}`, { params: { page, size } })
    return response.data
  },

  /** Get posts by a specific user */
  getPostsByUser: async (userId, page = 0, size = 10) => {
    const response = await apiClient.get(`/posts/user/${userId}`, { params: { page, size } })
    return response.data
  },
}

export default postService

File: backend/src/main/java/com/bloghub/api/service/PostService.java
package com.bloghub.api.service;

import com.bloghub.api.dto.PagedResponse;
import com.bloghub.api.dto.PostDto;
import com.bloghub.api.dto.PostRequest;

/**
 * Contract for blog post CRUD, pagination, search, and like operations.
 */
public interface PostService {

    PostDto createPost(PostRequest request, String username);

    PostDto updatePost(Long postId, PostRequest request, String username);

    void deletePost(Long postId, String username);

    PostDto getPostById(Long postId, String username);

    PagedResponse<PostDto> getAllPosts(int page, int size, String sortBy, String sortDir, String username);

    PagedResponse<PostDto> getPostsByUser(Long userId, int page, int size, String username);

    PagedResponse<PostDto> searchPosts(String query, int page, int size, String username);

    PagedResponse<PostDto> getPostsByCategory(String category, int page, int size, String username);

    PostDto toggleLike(Long postId, String username);
}
